internal/mtls/cert_utils.go

```diff
@@ -0,0 +1,174 @@
+package mtls
+
+import (
+	"bytes"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"fmt"
+	"math/big"
+	"net"
+	"time"
+)
+
+// CertificateGroup a bunch of methods to help to work with certificates.
+type CertificateGroup struct {
+	cert       *x509.Certificate
+	signedCert *x509.Certificate
+	privKey    *rsa.PrivateKey
+	certBytes  []byte
+	certPEM    *bytes.Buffer
+	PrivKeyPEM *bytes.Buffer
+}
+
+func NewCertificateGroupFromCACM(configMap map[string][]byte) (*CertificateGroup, error) {
+	certGroup := &CertificateGroup{
+		certPEM:    bytes.NewBuffer(configMap["ca.crt"]),
+		PrivKeyPEM: bytes.NewBuffer(configMap["ca.key"]),
+	}
+
+	block, _ := pem.Decode(certGroup.certPEM.Bytes())
+	if block == nil {
+		return nil, fmt.Errorf("Cannot get CA certificate")
+	}
+	ca, err := x509.ParseCertificate(block.Bytes)
+	if err != nil {
+		return nil, fmt.Errorf("Failing parsing cert: %v", err)
+	}
+	block, _ = pem.Decode(certGroup.PrivKeyPEM.Bytes())
+	if block == nil {
+		return nil, fmt.Errorf("Cannot get CA certificate key")
+	}
+
+	key, err := x509.ParsePKCS1PrivateKey(block.Bytes)
+	if err != nil {
+		return nil, fmt.Errorf("failing parsing key: %v", err)
+	}
+
+	certGroup.cert = ca // Not real at all, because this is already signed.
+	certGroup.signedCert = ca
+	certGroup.privKey = key
+	return certGroup, nil
+}
+
+// CreatePem from the load certificates create the PEM file and stores in local
+func (c *CertificateGroup) CreatePem() {
+
+	caPEM := new(bytes.Buffer)
+	pem.Encode(caPEM, &pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: c.certBytes,
+	})
+
+	c.certPEM = caPEM
+
+	privKeyPEM := new(bytes.Buffer)
+	pem.Encode(privKeyPEM, &pem.Block{
+		Type:  "RSA PRIVATE KEY",
+		Bytes: x509.MarshalPKCS1PrivateKey(c.privKey),
+	})
+	c.PrivKeyPEM = privKeyPEM
+}
+
+func (c *CertificateGroup) parseSignedCertificate() error {
+	var err error
+	c.signedCert, err = x509.ParseCertificate(c.certBytes)
+	return err
+}
+
+// GetCertificate returns the certificate Group in tls.Certficicate format.
+func (c *CertificateGroup) GetCertificate() (tls.Certificate, error) {
+	return tls.X509KeyPair(c.certPEM.Bytes(), c.PrivKeyPEM.Bytes())
+}
+
+func (c *CertificateGroup) GetCert() *x509.Certificate {
+	return c.cert
+}
+
+func getCACertificate() (*CertificateGroup, error) {
+	ca := &x509.Certificate{
+		SerialNumber: big.NewInt(time.Now().Unix()),
+		Subject: pkix.Name{
+			Organization: []string{"K4e-operator"},
+		},
+		NotBefore:             time.Now(),
+		NotAfter:              time.Now().AddDate(10, 0, 0),
```

Comment:
Isn't 10 years validity for the certificate too long ? Should the certificate period be configurable and set based on org policy ?

Reply:
This is only for CA

Reply:
How do we rotate CA when 10y passes? We can check validation period and regenerate and we would need to force agents to reregister.

Reply:
We can follow the same way as here: 
https://www.cockroachlabs.com/docs/stable/rotate-certificates.html

Reply:
Is this blocking the PR, should I implement the process to Update the CA cert?

Reply:
Issue to track this: 
https://issues.redhat.com/browse/ECOPROJECT-432